CoreFoundation/NumberDate.subproj/CFTimeZone.c

```diff
@@ -183,11 +186,178 @@ static void __CFTimeZoneGetOffset(CFStringRef timezone, int32_t *offset) {
 
     RegCloseKey(hKey);
 }
+#elif TARGET_OS_ANDROID
+/*
+ * Android does not ship with the standard Unix Olsen files, with the directory
+ * structure, and the files with the same name as the time zone.
+ * Instead all the information is in one file, where all the time zone
+ * information for all the zones is held. Also, to allow upgrades to the time
+ * zone database without an update to the system, the database can be overriden
+ * by a secondary location.
+ * - /data/misc/zoneinfo/current/tzdata overrides for the time zone information
+ *   from the system.
+ * - /system/usr/share/zoneinfo/tzdata system time zone information.
+ * The format of these files is slightly documented in Bionic's source file
+ * libc/tzcode/bionic.cpp.
+ * The file start with a header which is 24 bytes long.
+ * - 6 bytes should be the ASCII string "tzdata"
+ * - 6 bytes of the Olson database version in ASCII. For example 2018a. Includes
+ *   a final nul character.
+ * - 4 bytes MSB of the offset of the index inside the file.
+ * - 4 bytes MSB of the offset of the start of the data inside the file.
+ * - 4 bytes MSB of the zonetab (unused in this code).
+ * The index sits between the offset for the index and the offset for the data.
+ * Each index entry is 52 bytes long.
+ * - 40 bytes for the name of the zone. Seems to be nul terminated.
+ * - 4 bytes MSB of the offset to this time zone data. Notice that this offset
+ *   is relative to the data offset from the header.
+ * - 4 bytes MSB of the data length.
+ * - 4 bytes unused.
+ */
+
+#define ANDROID_TZ_HEADER_SIZE 24
+#define ANDROID_TZ_HEADER_TAG_SIZE 6
+#define ANDROID_TZ_HEADER_INDEX_OFFSET 12
+#define ANDROID_TZ_HEADER_DATA_OFFSET 16
+#define ANDROID_TZ_ENTRY_SIZE 52
+#define ANDROID_TZ_ENTRY_NAME_LENGTH 40
+#define ANDROID_TZ_ENTRY_START_OFFSET 40
+#define ANDROID_TZ_ENTRY_LENGTH_OFFSET 44
+
+/**
+ * Callback invoked for each of the time zones in the timezone file.
+ * - name: The name of the time zone.
+ * - offset: The final offset inside the file of the data for the time zone.
+ * - length: The length of the data for the time zone.
+ * - fp: The file pointer to read the data from. The file might not be in the
+ *       right offset, so if you want to read the right time zone data, you
+ *       probably want to fseek into offset.
+ * - context1: The argument passed into __CFAndroidTimeZoneListEnumerate.
+ * - context2: The argument passed into __CFAndroidTimeZoneListEnumerate.
+ */
+typedef Boolean (*__CFAndroidTimeZoneListEnumerateCallback)(const char name[ANDROID_TZ_ENTRY_NAME_LENGTH], int32_t offset, int32_t length, FILE *fp, void *context1, void *context2);
```

Comment:
I would give them proper names rather than `context1` and `context2`.  (e.g. `name`, `data`)

Reply:
They are `name` and `data` in the case you are calling `__CFTimeZoneDataCreateCallback`, but `context1` is actually a `CFMutableArrayRef` in the case of calling `__CFCopyAndroidTimeZoneListCallback`. That's why I used just one `context` before.